### PR TITLE
Add `fence`, `cmpxchg` and `atomicRMW` to the builder

### PR DESCRIPTION
**Pull request by CodaFi**
The next instructions I'm going to focus on will be the `extract*` family because `atomic cmpxchg` is pretty useless without it.

### CODE REVIEW

**Comment by CodaFi** [comment]
With this patch the builder is quickly approaching 1500 LoC.  It's gonna need to be broken up soon.

File: Sources/LLVM/IRBuilder.swift [hunk 1]
```diff
@@ -103,13 +105,196 @@ public enum RealPredicate {
...
+  /// only reads, release for an operation that only writes), there is a global 
+  /// total order on all sequentially-consistent operations on all addresses, 
+  /// which is consistent with the happens-before partial order and with the 
+  /// modification orders of all the affected addresses. Each 
+  /// sequentially-consistent read sees the last preceding write to the same 
+  /// address in this global order. This corresponds to the C++0x/C1x 
+  /// `memory_order_seq_cst` and Java volatile.
+  case sequentiallyConsistent = 6
+
+  static let orderingMapping: [AtomicOrdering: LLVMAtomicOrdering] = [
```
Comment (harlanhaskins) [inline Sources/LLVM/IRBuilder.swift]: Private?

File: Sources/LLVM/IRBuilder.swift [hunk 2]
```diff
@@ -103,13 +105,196 @@ public enum RealPredicate {
...
+
+  /// Retrieves the corresponding `LLVMAtomicOrdering`.
+  public var llvm: LLVMAtomicOrdering {
+    return AtomicOrdering.orderingMapping[self]!
+  }
+}
+
+/// `AtomicRMWOperation` enumerates the kinds of supported atomic
+/// read-write-modify operations.
+public enum AtomicRMWOperation {
```
Comment (harlanhaskins) [inline Sources/LLVM/IRBuilder.swift]: Maybe spell this out.

File: Sources/LLVM/IRBuilder.swift [hunk 3]
```diff
@@ -942,6 +1127,91 @@ public class IRBuilder {
...
+  ///
+  /// - returns: A value representing the original value at the given location 
+  ///   is together with a flag indicating success (true) or failure (false).
+  public func buildAtomicCmpXchg(
+    ptr: IRValue, of old: IRValue, to new: IRValue,
+    successOrdering: AtomicOrdering, failureOrdering: AtomicOrdering,
+    singleThreaded: Bool = false
+  ) -> IRValue {
+
+    if failureOrdering.rawValue < AtomicOrdering.monotonic.rawValue {
```
Comment (harlanhaskins) [inline Sources/LLVM/IRBuilder.swift]: Maybe just make a conformance to Comparable?

**Comment by CodaFi** [comment]
@harlanhaskins ping

**Review by harlanhaskins — APPROVED** [review]
:shipit:

**Pull request merged**